src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,241 @@
+                .authenticatorSelection(authenticatorSelection).timeout(TIMEOUT).build();
+        final PublicKeyCredentialCreationOptions options = rp.startRegistration(startRegistrationOptions);
+        final WebAuthnCredentialCreationInfo info = new WebAuthnCredentialCreationInfo();
+        info.username = username;
+        info.realm = realm;
+        info.options = options;
+        info.providerId = provider;
+        final String key = generateNewKey();
+        activeRegistrations.put(key, info);
+        return Pair.of(options, key);
+    }
+
+    /**
+     * Returns the username of the user on successful authentication or null if the
+     * authentication was not successful
+     */
+    public Optional<String> finishRegistration(
+            PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc,
+            String realm,
+            String key) {
+        try {
+            final WebAuthnCredentialCreationInfo info = activeRegistrations.get(key);
+            if (info == null || info.realm != realm) {
+                return Optional.empty();
+            }
+            final String username = info.username;
+            final WebAuthnUserAccount account = webAuthnUserAccountRepository.findByProviderAndUsername(provider,
+                    username);
+            assert (account != null);
+            final PublicKeyCredentialCreationOptions options = info.options;
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
+                final Set<WebAuthnCredential> previousCredentials = account.getCredentials();
+                final WebAuthnCredential newCred = new WebAuthnCredential();
+                newCred.setCreatedOn(new Date());
+                newCred.setLastUsedOn(new Date());
+                newCred.setCredentialId(result.getKeyId().getId());
+                newCred.setPublicKeyCose(result.getPublicKeyCose());
+                newCred.setSignatureCount(result.getSignatureCount());
+                newCred.setTransports(result.getKeyId().getTransports().orElse(new TreeSet<>()));
+                newCred.setParentAccount(account);
+
+                previousCredentials.add(newCred);
+                webAuthnCredentialsRepository.save(newCred);
+                account.setCredentials(previousCredentials);
+                webAuthnUserAccountRepository.save(account);
+                activeRegistrations.remove(key);
+                return Optional.of(username);
+            }
+        } catch (Exception e) {
+            System.out.println(e);
+        }
+        return Optional.empty();
+    }
+
+    public Pair<AssertionRequest, String> startLogin(String username, String realm) {
```

Comment:
needs refactoring, return object is not well defined

Reply:
Redefined return object to existing DTO, `WebAuthnLoginResponse`, in f5dda8a98fb396dccffe3f21455aa6b931bbd3f6